File: 0994-rotting-oranges/0994-rotting-oranges.java
class Solution {
    public int orangesRotting(int[][] grid) {
        int row = grid.length;
        int col = grid[0].length;
        int freshCount = 0;
        Queue<int[]> queue = new LinkedList<>();

        for (int r = 0; r < row; r++){
            for (int c = 0; c < col; c++){
                if (grid[r][c] == 2){
                    queue.offer(new int[]{r,c});
                }else if (grid[r][c] == 1){
                    freshCount++;
                }
            }
        }

        if (freshCount ==0){
            return 0;
        }
        
        int[][] directions = {{-1,0}, {1,0}, {0,1}, {0,-1}};
        int minutes = 0;

        while (!queue.isEmpty() && freshCount > 0){
            int levelSize = queue.size();
            boolean hasNewRotten = false;

            for (int i = 0; i < levelSize; i++){
                int[] current = queue.poll();
                int curr_r = current[0];
                int curr_c = current[1];

                for (int[] dir:directions){
                    int newRow = curr_r + dir[0];
                    int newCol = curr_c + dir[1];

                    if (0 <= newRow && newRow < row &&
                    0 <=newCol && newCol< col &&
                    grid[newRow][newCol] == 1){
                        grid[newRow][newCol] = 2;
                        freshCount--;
                        queue.offer(new int[]{newRow, newCol});
                        hasNewRotten = true;
                    }
                }
            }
            if (hasNewRotten){
                minutes += 1;
            }
        }
        return freshCount == 0 ? minutes : -1;

    }
}

File: 0994-rotting-oranges/0994-rotting-oranges.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        row = len(grid)
        col = len(grid[0])
        queue = deque([])
        elapse = 0
        for i in range(row):
            for j in range(col):
                if grid[i][j] == 2:
                    queue.append([i,j])
        directions = [[-1,0],[1,0],[0,1],[0,-1]]
        while queue:
            q = len(queue)
            newRotten = False
            for i in range(q):
                x,y = queue.popleft()
                
                for dx,dy in directions:
                    new_x = x+dx
                    new_y = y + dy
                    if 0 <= new_x < row and 0 <= new_y < col and grid[new_x][new_y]==1:
                        newRotten = True
                        grid[new_x][new_y] = 2
                        queue.append([new_x, new_y])
            if newRotten:
                elapse += 1
        for i in range(row):
            for j in range(col):
                if grid[i][j] == 1:
                    return -1
        return elapse    

                

        